File: 数据结构和算法/AVLTree.java
import java.util.LinkedList;
import java.util.Queue;

/**
 * 平衡二叉树（AVL）
 */
public class AVLTree<T extends Comparable<T>> {

  // 根节点
  private AVLNode<T> root;

  /**
   * 右旋（LL型）
   * 
   * @param node 不平衡子树根节点节点
   * @return 旋转后的子树
   */
  public AVLNode<T> singleRotateRight(AVLNode<T> node) {
    AVLNode<T> leftNode = node.left;
    node.left = leftNode.right;
    leftNode.right = node;

    // 更新高度
    node.height = Math.max(height(node.left), height(node.right)) + 1;
    leftNode.height = Math.max(height(leftNode.left), node.height) + 1;

    return leftNode;
  }

  /**
   * 右旋（RR型）
   * 
   * @param node
   * @return
   */
  public AVLNode<T> singleRotateLeft(AVLNode<T> node) {
    AVLNode<T> rightNode = node.right;
    node.right = rightNode.left;
    rightNode.left = node;

    // 更新高度
    node.height = Math.max(height(node.left), height(node.right)) + 1;
    rightNode.height = Math.max(height(rightNode.right), node.height) + 1;

    return rightNode;
  }

  /**
   * 左-右旋（LR型）
   * 
   * @param node
   * @return
   */
  public AVLNode<T> doubleRotateLeftRight(AVLNode<T> node) {
    // 先对左子树进行左旋
    node.left = this.singleRotateLeft(node.left);
    // 对不平衡节点进行右旋
    return this.singleRotateRight(node);
  }

  /**
   * 右-左旋（RL型）
   * 
   * @param node
   * @return
   */
  public AVLNode<T> doubleRotateRightLeft(AVLNode<T> node) {
    // 先对其右子树进行右旋
    node.right = this.singleRotateRight(node.right);
    // 对不平衡节点进行左旋
    return this.singleRotateLeft(node);
  }

  public int height() {
    return this.height(this.root);
  }

  public int height(AVLNode<T> node) {
    return node == null ? -1 : node.height;
  }

  public void insert(T data) {
    this.root = this._insert(root, data);
  }

  private AVLNode<T> _insert(AVLNode<T> node, T data) {
    if (node == null) { // 创建新节点
      node = new AVLNode<T>(data);
    }

    // 往左添加
    if (data.compareTo(node.data) < 0) {

      node.left = _insert(node.left, data);

      // 是否失衡
      if (height(node.left) - height(node.right) == 2) {
        // 判断是插入失衡点左孩子的左子树还是右子树
        if (data.compareTo(node.left.data) < 0) {
          // LL型
          node = singleRotateRight(node);
        } else {
          // LR型
          node = doubleRotateLeftRight(node);
        }
      }

      // 往右添加
    } else if (data.compareTo(node.data) > 0) {

      node.right = _insert(node.right, data);

      // 判断是否失衡
      if (height(node.right) - height(node.left) == 2) {
        if (data.compareTo(node.right.data) < 0) {
          // RL型
          node = doubleRotateRightLeft(node);
        } else {
          // RR型
          node = singleRotateLeft(node);
        }
      }
    }

    node.height = Math.max(height(node.left), height(node.right)) + 1;

    return node;
  }

  /**
   * 删除方法
   * 
   * @param data
   */
  public void remove(T data) {
    if (data == null) {
      throw new RuntimeException("data can\'t not be null ");
    }
    this.root = remove(data, root);
  }

  /**
   * 删除操作
   * 
   * @param data
   * @param p
   * @return
   */
  private AVLNode<T> remove(T data, AVLNode<T> p) {

    if (p == null)
      return null;

    int result = data.compareTo(p.data);

    // 从左子树查找需要删除的元素
    if (result < 0) {
      p.left = remove(data, p.left);

      // 检测是否平衡
      if (height(p.right) - height(p.left) == 2) {
        AVLNode<T> currentNode = p.right;
        // 判断需要那种旋转
        if (height(currentNode.right) >= height(currentNode.left)) {
          // RR
          p = singleRotateLeft(p);
        } else {
          // RL
          p = doubleRotateRightLeft(p);
        }
      }

    }
    // 从右子树查找需要删除的元素
    else if (result > 0) {
      p.right = remove(data, p.right);
      // 检测是否平衡
      if (height(p.left) - height(p.right) == 2) {
        AVLNode<T> currentNode = p.left;
        // 判断需要那种旋转
        if (height(currentNode.left) >= height(currentNode.right)) {
          // LL
          p = singleRotateRight(p);
        } else {
          // LR
          p = doubleRotateLeftRight(p);
        }
      }
    }
    // 已找到需要删除的元素,并且要删除的结点拥有两个子节点
    else if (p.right != null && p.left != null) {

      // 寻找替换结点
      p.data = findMin(p.right).data;

      // 移除用于替换的结点
      p.right = remove(p.data, p.right);
    } else {
      // 只有一个孩子结点或者只是叶子结点的情况
      p = (p.left != null) ? p.left : p.right;
    }

    // 更新高度值
    if (p != null)
      p.height = Math.max(height(p.left), height(p.right)) + 1;
    return p;
  }

  public AVLNode<T> findMin(AVLNode<T> node) {
    if (node == null)
      return null;
    if (node.left == null)
      return node;

    return findMin(node.left);
  }

  /**
   * 使用队列进行层级遍历
   */
  public void levelOrderTranverse() {
    Queue<AVLNode<T>> queue = new LinkedList<AVLNode<T>>();

    queue.offer(this.root);

    while (!queue.isEmpty()) {
      AVLNode<T> node = queue.poll();
      if (node.left != null) {
        queue.offer(node.left);
      }
      if (node.right != null) {
        queue.offer(node.right);
      }
      System.out.print(node.data + ", ");
    }
  }

  public static void main(String[] args) {
    AVLTree<Integer> tree = new AVLTree<Integer>();
    int[] arr = new int[] { 50, 45, 60, 30, 46, 47, 55, 70, 80 };
    for (int i = 0; i < arr.length; i++) {
      tree.insert(arr[i]);
    }

    tree.levelOrderTranverse();
    tree.remove(80);
    System.out.println("");
    tree.levelOrderTranverse();
    System.out.println(tree.height());

  }

}

/**
 * 节点对象
 * 
 * @param <T> 节点类型
 */
class AVLNode<T> {
  public AVLNode<T> left;
  public AVLNode<T> right;
  public int height;
  public T data;

  public AVLNode(T data) {
    this.data = data;
  }

  public AVLNode(T data, AVLNode<T> left, AVLNode<T> right) {
    this.data = data;
    this.left = left;
    this.right = right;
  }
}

File: 数据结构和算法/AVLTree.ts
/**
 * 使用ts实现AVL
 */
class AVLTree<T> {
  private root: AVLNode<T>

  constructor() {
    this.root = null
  }

  // LL型
  singleRotateRight(node: AVLNode<T>): AVLNode<T> {
    const leftNode = node.left
    node.left = leftNode.right
    leftNode.right = node

    // 更新高度
    node.height = Math.max(this.height(node.left), this.height(node.right)) + 1
    leftNode.height = Math.max(this.height(leftNode.left), node.height) + 1
    return leftNode
  }

  // RR型
  singleRotateLeft(node: AVLNode<T>): AVLNode<T> {
    const rightNode = node.right
    node.right = rightNode.left
    rightNode.left = node

    // 更新高度
    node.height = Math.max(this.height(node.left), this.height(node.right)) + 1
    rightNode.height = Math.max(this.height(rightNode.right), node.height) + 1
    return rightNode
  }

  // LR型
  doubleRotateLeftRight(node: AVLNode<T>): AVLNode<T> {
    node.left = this.singleRotateLeft(node.left)
    return this.singleRotateRight(node)
  }

  // RL型
  doubleRotateRightLeft(node: AVLNode<T>): AVLNode<T> {
    node.right = this.singleRotateRight(node.right)
    return this.singleRotateLeft(node)
  }

  insert(data: T) {
    this.root = this._insert(this.root, data)
  }

  _insert(node: AVLNode<T>, data: T): AVLNode<T> {
    if (node === null) {
      node = new AVLNode<T>(data)
      return node
    }

    if (data < node.data) {
      node.left = this._insert(node.left, data)
      // 判断是否失衡
      if (this.height(node.left) - this.height(node.right) === 2) {
        if (data < node.left.data) {
          node = this.singleRotateRight(node) // LL
        } else {
          node = this.doubleRotateLeftRight(node) // LR
        }
      }
    } else {
      node.right = this._insert(node.right, data)
      // 判断是否失衡
      if (this.height(node.right) - this.height(node.left) === 2) {
        if (data < node.right.data) {
          node = this.doubleRotateRightLeft(node) // RL
        } else {
          node = this.singleRotateLeft(node) //RR
        }
      }
    }

    // 调整高度
    node.height = Math.max(this.height(node.left), this.height(node.right)) + 1

    return node
  }

  heightRoot(): number {
    return this.height(this.root);
  }

  height(node: AVLNode<T>): number {
    return node === null ? -1 : node.height
  }

  remove(data: T) {
    if (!data) return;
    this.root = this._remove(this.root, data)
  }

  _remove(node: AVLNode<T>, data: T): AVLNode<T> {
    if (node === null) return null

    if (data < node.data) {
      node.left = this._remove(node.left, data)
      // 判断是否平衡
      if (this.height(node.right) - this.height(node.left) === 2) {
        const rightNode = node.right
        if (this.height(rightNode.left) <= this.height(rightNode.right)) { // RR
          node = this.singleRotateLeft(node)
        } else { //RL
          node = this.doubleRotateRightLeft(node)
        }
      }

    } else if (data > node.data) {
      node.right = this._remove(node.right, data)
      // 判断是否平衡
      if (this.height(node.left) - this.height(node.right) === 2) {
        const leftNode = node.left
        if (this.height(leftNode.left) >= this.height(leftNode.right)) { // LL
          node = this.singleRotateRight(node)
        } else { // LR
          node = this.doubleRotateLeftRight(node)
        }
      }

    } else { // 找到了被删除节点
      if (node.right !== null && node.left !== null) {
        node.data = this.findMin(node.right).data
        node.right = this._remove(node.right, node.data)
      } else {
        node = (node.left !== null) ? node.left : node.right
      }
    }

    // 更新高度
    if (node !== null) {
      node.height = Math.max(this.height(node.left), this.height(node.right)) + 1
    }
    return node
  }

  findMin(node: AVLNode<T>): AVLNode<T> {
    if (node === null) return null
    if (node.left === null) return node
    return this.findMin(node.left)
  }

  levelOrderTranverse(cb) {
    const queue = []
    queue.push(this.root)
    while (queue.length > 0) {
      const cur: AVLNode<T> = queue.shift()
      if (cur.left) {
        queue.push(cur.left)
      }
      if (cur.right) {
        queue.push(cur.right)
      }
      cb && cb(cur.data)
    }
  }
}


class AVLNode<T> {
  left: AVLNode<T>
  right: AVLNode<T>
  height: number
  data: T

  constructor(data: T) {
    this.data = data
    this.left = null;
    this.right = null;
    this.height = 0;
  }
}


// 测试
const arr = [50, 45, 60, 30, 46, 47, 55, 70, 80, 20, 20, 19]
const tree = new AVLTree<number>()
arr.forEach(item => {
  tree.insert(item)
})
tree.remove(55)
tree.levelOrderTranverse(item => console.log(item))
